src/psyclone/tests/psyGen_test.py

```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+
+    # Test properly formed but empty kernel module
+    schedule = processor.generate_schedule("dummy_code", ast2)
+    assert isinstance(schedule, KernelSchedule)
+
+    # Test generate an unexistent subroutine name
+    with pytest.raises(InternalError) as error:
+        schedule = processor.generate_schedule("nonexistent_code", ast2)
+    assert "Unexpected kernel AST. Could not find " \
+           "subroutine: nonexistent_code" in str(error.value)
+
+    # Test corrupting ast by deleting subroutine
+    del ast2.content[0].content[2]
+    with pytest.raises(InternalError) as error:
+        schedule = processor.generate_schedule("dummy_code", ast2)
+    assert "Unexpected kernel AST. Could not find " \
+           "subroutine: dummy_code" in str(error.value)
+
+    # Test corrupting ast by deleting specification part
+    del ast2.content[0].content[1]
+    with pytest.raises(InternalError) as error:
+        schedule = processor.generate_schedule("dummy_code", ast2)
+    assert "Unexpected kernel AST. Could not find specification part."
+
+    DUMMY_KERNEL_METADATA = '''
+    module dummy_mod
+      type, extends(kernel_type) :: dummy_type
+         type(arg_type), meta_args(3) =                    &
+              (/ arg_type(gh_field, gh_write,     w3),     &
+                 arg_type(gh_field, gh_readwrite, wtheta), &
+                 arg_type(gh_field, gh_inc,       w1)      &
+               /)
+         integer :: iterates_over = cells
+       contains
+         procedure, nopass :: code => dummy_code
+      end type dummy_type
+    contains
+     subroutine dummy_code(f1, f2, f3)
+        real(wp), dimension(:,:), intent(in)  :: f1
+        real(wp), dimension(:,:), intent(out)  :: f2
+        real(wp), dimension(:,:), intent(inout)  :: f3
+        f2 = f1 + 1
+      end subroutine dummy_code
+    end module dummy_mod
+    '''
+    ast1 = fpapi.parse(DUMMY_KERNEL_METADATA, ignore_comments=True)
+    metadata = DynKernMetadata(ast1)
+    my_kern = DynKern()
+    my_kern.load_meta(metadata)
+    ast2 = my_kern.ast
+
+    # Test properly formed kernel module
+    schedule = processor.generate_schedule("dummy_code", ast2)
+    assert isinstance(schedule, KernelSchedule)
+
+    # Test corrupting ast by deleting execution part
+    del ast2.content[0].content[2].content[1].content[2]
+    schedule = processor.generate_schedule("dummy_code", ast2)
+    assert isinstance(schedule, KernelSchedule)
+
+
+def test_fparser2astprocessor_process_declarations():
+    from fparser.two.parser import ParserFactory
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Specification_Part
+    ParserFactory().create(std="f2008")
+    fake_parent = KernelSchedule("dummy_schedule")
+    processor = Fparser2ASTProcessor()
+
+    # Test simple declarations
+    reader = FortranStringReader("integer :: l1")
+    fparser2specification = Specification_Part.match(reader)[0][0]
```

Comment:
Can you just call the constructor rather than having to go via the match method?

Reply:
I think so. You can do:

    fparser2spec = Specification_Part(reader).content[0]

Reply:
Also I didn't get your first comment here, indeed the proposed modification using the constructor looks nicer. I made the change throughout the new unit tests.